Add a validate-vendor script

vdemeester opened the pull request:
Makes sure that if ./hack/vendor.sh has been updated, it has been used to update the vendor folder. 🐢

This is a complement to #19141 in order to keep the `vendor/` folder in sync with `./hack/vendor.sh`.

``` bash
# with a wrong vendor.sh
$ make validate
# […]
done
These files are not valid vendor file:
 - hack/vendor.sh

Please vendor your package with ./hack/vendor.sh.

```
- The message in case of failure <del>might</del> definitely need some _lifting_ :wink:.
- This makes a build with updates on vendoring a little longer.
- Should we run this script all the time for the build on the `master` branch or not ? 

/cc @tianon @jfrazelle :angel: 

_There is a commit to trigger the failure. I'll remove it when everything is good and validated :stuck_out_tongue_closed_eyes:._

🐸

Build failure example : https://jenkins.dockerproject.org/job/Docker-PRs/22003/console

``` bash
08:15:00 ---> Making bundle: validate-vendor (in bundles/1.10.0-dev/validate-vendor)

08:17:38 The result of ./hack/vendor.sh differs
08:17:38 
08:17:38  M vendor/src/github.com/mistifyio/go-zfs/CONTRIBUTING.md
08:17:38  M vendor/src/github.com/mistifyio/go-zfs/utils.go
08:17:38  M vendor/src/github.com/mistifyio/go-zfs/zfs.go
08:17:38  M vendor/src/github.com/mistifyio/go-zfs/zpool.go
08:17:38 
08:17:38 Please vendor your package with ./hack/vendor.sh.
08:17:38 

08:17:38 Build step 'Execute shell' marked build as failure
```

Signed-off-by: Vincent Demeester vincent@sbr.pm

vdemeester commented:
For a build failure, look at https://jenkins.dockerproject.org/job/Docker-PRs/21963/console :wink:

tianon commented:
Neat!  Maybe pipe the output of running `hack/vendor.sh` to `/dev/null` as well so we can still read the output of the validate script nicely?

I'd thought about doing this earlier, but was a little worried because it _is_ destructive to the local copy and I couldn't think of a clean way to avoid that but for this script it's probably not a big deal (having bad `vendor` changes is definitely worse). :+1:

vdemeester commented:
@tianon true, it's a little too much output right now :sweat_smile:.

tianon commented on hack/make/validate-vendor:
```diff
@@ -0,0 +1,33 @@
+source "${MAKEDIR}/.validate"
+
+IFS=$'\n'
+files=( $(validate_diff --diff-filter=ACMR --name-only -- 'hack/vendor.sh' || true) )
```
We probably want to adjust this so that it includes `vendor/` as well, and then change the `for` loop below to instead be something like `if [ "${#files[@]}" -gt 0 ]; then` so that it runs if either `hack/vendor.sh` or anything in the `vendor` directory have changed and verifies both.

vdemeester replied:
:heart_eyes_cat:

tianon commented on hack/make/validate-vendor:
```diff
@@ -0,0 +1,33 @@
+	# We run vendor.sh to and see if we have a diff afterwards
+	./hack/vendor.sh
+	# Let see if the working directory is clean
+	diffs=( $(git status --porcelain 2>/dev/null| egrep "^( M|A|D)") )
```
hmm, this is actually going to have false-positives if we've got other changes in our local copy -- maybe just `git status --porcelain -- vendor` to make sure it's limited to just the vendor directory's changes?  Then, you could store it in a string instead, and check whether the string is empty, and then use the string as the error message output telling us where the problems are. :smile:

tianon replied:
ie, something like:
```bash
diff="$(git status --porcelain -- vendor 2>/dev/null)"
if [ "$diff" ]; then
  # bad juju, store the diff for display
fi
```

vdemeester replied:
ahh you are right about that :+1:

duglin commented on hack/make/validate-vendor:
```diff
@@ -0,0 +1,26 @@
+			echo
+			echo 'Please vendor your package with ./hack/vendor.sh.'
+			echo
+		}
```
need an `exit 1` or something so its non-zero

duglin commented on hack/make/validate-vendor:
```diff
@@ -0,0 +1,26 @@
+		{
+			echo 'The result of ./hack/vendor.sh differs'
+			echo
+			echo $diffs
```
you may want to tweak the formatting of this so that each file appears on its own line.
right now it shows as:
```
23:01:35 ---> Making bundle: validate-vendor (in bundles/1.10.0-dev/validate-vendor)

23:04:11 The result of ./hack/vendor.sh differs
23:04:11 
23:04:11 M vendor/src/github.com/mistifyio/go-zfs/CONTRIBUTING.md M vendor/src/github.com/mistifyio/go-zfs/utils.go M vendor/src/github.com/mistifyio/go-zfs/zfs.go M vendor/src/github.com/mistifyio/go-zfs/zpool.go
23:04:11 
23:04:11 Please vendor your package with ./hack/vendor.sh.

```

tianon replied:
Quotes around "$diffs" should fix that problem. :+1:

duglin commented:
Output at: https://jenkins.dockerproject.org/job/Docker-PRs/22003/console looks good to me!
so LGTM once we remove the extra commit.

duglin commented on hack/make/validate-vendor:
```diff
@@ -0,0 +1,27 @@
+source "${MAKEDIR}/.validate"
+
+IFS=$'\n'
+files=( $(validate_diff --diff-filter=ACMR --name-only -- 'hack/vendor.sh' 'vendor/' || true) ) 
```
should this also look for "hack/.vendor-helpers.sh" ?

vdemeester replied:
@duglin good point :+1:

vdemeester committed:
Add a validate-vendor script

Makes sure that if ./hack/vendor.sh has been updated, it has been used
to update the vendor folder.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

duglin commented:
still LGTM
ping @tianon

tianon commented:
Niiiiiiiice! :smile: 

LGTM

jfrazelle commented:
LGTM
